### PR TITLE
MODROLESKC-214: permission replacement support

### PR DESCRIPTION
**Pull request by mykolam-epm**
## Purpose

Support for permission "replaces" feature that allows one to replace old permissions with new ones.

## Pre-Merge Checklist:

Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added, or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do Rally stories exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail? Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the Rally stories under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally, all the PRs involved in breaking changes would be merged on the same day to avoid breaking the folio-testing
environment. Communication is paramount if that is to be achieved, especially as the number of inter-module and
inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the
responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/roles/domain/model/CapabilityReplacements.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package org.folio.roles.domain.model;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+public record CapabilityReplacements(Map<String, Set<String>> oldCapabilitiesToNewCapabilities,
+                                     Map<String, Set<UUID>> oldCapabilityRoleAssignments,
```
Comment (OleksiiKuzminov) [inline src/main/java/org/folio/roles/domain/model/CapabilityReplacements.java:8]: format

File: src/main/java/org/folio/roles/integration/kafka/CapabilityEventProcessor.java [hunk 2]
```diff
@@ -99,7 +100,7 @@ private CapabilityResultHolder processUiModuleResources(CapabilityEvent event, L
   private Optional<CapabilitySetDescriptor> createCapabilitySetDescriptor(
     CapabilityEvent event, FolioResource resource, ModuleType moduleType) {
     var folioPermission = resource.getPermission().getPermissionName();
-    var permissionMappingOverrides = emptyIfNull(event.getPermissionMappingOverrides());
+    var permissionMappingOverrides = permissionOverrider.getPermissionMappings();
```
Comment (OleksiiKuzminov) [inline src/main/java/org/folio/roles/integration/kafka/CapabilityEventProcessor.java:103]: cant be null?

**Comment by OleksiiKuzminov** [comment]
@mykolam-epm It is a very complex functionality, please add integration tests. Without these tests, we can't be sure that it works properly and that future changes do not break something.

File: src/main/java/org/folio/roles/integration/kafka/KafkaMessageListener.java [hunk 3]
```diff
@@ -31,8 +29,10 @@ public class KafkaMessageListener {
     groupId = "#{folioKafkaProperties.listener['capability'].groupId}",
     topicPattern = "#{folioKafkaProperties.listener['capability'].topicPattern}")
   public void handleCapabilityEvent(ResourceEvent resourceEvent) {
-    try (var ignored = new FolioExecutionContextSetter(metadata, Map.of(TENANT, List.of(resourceEvent.getTenant())))) {
-      capabilityKafkaEventHandler.handleEvent(resourceEvent);
+    try (
+      var ignored = new FolioExecutionContextSetter(executionContextBuilder.buildContext(resourceEvent.getTenant()))) {
```
Comment (yauhen-vavilkin) [inline src/main/java/org/folio/roles/integration/kafka/KafkaMessageListener.java:33]: It looks like the system user should be defined in MD. So, that place can be called as `executeSystemUserScoped`

File: pom.xml [hunk 4]
```diff
@@ -79,6 +79,12 @@
       <version>${folio-spring-support.version}</version>
     </dependency>
 
+    <dependency>
+      <groupId>org.folio</groupId>
+      <artifactId>folio-spring-system-user</artifactId>
```
Comment (yauhen-vavilkin) [inline pom.xml:84]: By default, the library creates SystemUserService, which is a legacy approach that doesn't work in Eureka. At least the system user default behavior should be overridden (`system-user.enabled = false`),  Ideally, the system user should also be defined.

File: src/main/java/org/folio/roles/service/permission/PermissionOverrider.java [hunk 5]
```diff
@@ -26,6 +27,7 @@ public class PermissionOverrider {
   private final ResourceLoader resourceLoader;
   private final PermissionMappingProperties permissionMappingProperties;
 
+  @Cacheable(cacheNames = "permission-mappings")
```
Comment (yauhen-vavilkin) [inline src/main/java/org/folio/roles/service/permission/PermissionOverrider.java:30]: To keep this functionality "dynamic", we have to configure cache TTL

File: src/test/resources/application-it.yml [hunk 6]
```diff
@@ -39,3 +39,9 @@ header:
     x-okapi-user-id:
       exclude:
         base-paths: /tests, /admin
+folio:
+  environment: test
+  system-user:
```
Comment (yauhen-vavilkin) [inline src/test/resources/application-it.yml:44]: folio.system-user.enabled: false

File: src/main/java/org/folio/roles/integration/kafka/KafkaMessageListener.java (hunk 3 above)
Comment (mykolam-epm) [reply src/main/java/org/folio/roles/integration/kafka/KafkaMessageListener.java:33]: What system user?

File: src/main/java/org/folio/roles/service/permission/PermissionOverrider.java (hunk 5 above)
Comment (mykolam-epm) [reply src/main/java/org/folio/roles/service/permission/PermissionOverrider.java:30]: Permission mappings never change in runtime - one needs to change a file that is packed inside the app JAR to change permission mappings, so it cannot be dynamic anyway.

File: src/main/java/org/folio/roles/integration/kafka/CapabilityEventProcessor.java (hunk 2 above)
Comment (mykolam-epm) [reply src/main/java/org/folio/roles/integration/kafka/CapabilityEventProcessor.java:103]: No, PermissionOverrider either returns result of `return permissions.entrySet().stream().collect(toMap(Map.Entry::getKey, PermissionOverrider::mapToPermissionData));` or throws exception.

File: src/main/java/org/folio/roles/domain/model/CapabilityReplacements.java (hunk 1 above)
Comment (mykolam-epm) [reply src/main/java/org/folio/roles/domain/model/CapabilityReplacements.java]: I'll change it here manually, but this is actually how our formatter config reformats records.
And this is a new code, not an existing one - so it's not that I changed existing formatting. FYI.

File: src/main/java/org/folio/roles/service/permission/PermissionOverrider.java (hunk 5 above)
Comment (mykolam-epm) [reply src/main/java/org/folio/roles/service/permission/PermissionOverrider.java:30]: Added cache config.

File: pom.xml (hunk 4 above)
Comment (mykolam-epm) [reply pom.xml:84]: Added definition of system user to module descriptor.

File: descriptors/ModuleDescriptor-template.json [hunk 7]
```diff
@@ -712,5 +712,14 @@
       { "name": "KC_TOKEN_TTL", "value": "59" },
       { "name": "KC_CLIENT_ID_TTL", "value": "3600" }
     ]
+  },
+  "metadata": {
+    "user": {
+      "type": "system",
+      "permissions": [
+  		"users-keycloak.auth-users.item.get",
```
Comment (yaroslav-kiriak) [inline descriptors/ModuleDescriptor-template.json:720]: formatting at #L720, #L722

Comment (mykolam-epm) [reply descriptors/ModuleDescriptor-template.json]: Done

File: src/main/java/org/folio/roles/integration/kafka/KafkaMessageListener.java (hunk 3 above)
Comment (mykolam-epm) [reply src/main/java/org/folio/roles/integration/kafka/KafkaMessageListener.java:33]: Done

**Comment by mykolam-epm** [comment]
Added integration test that verifies replacement of permissions and their assignments.

File: src/main/resources/application.yml [hunk 8]
```diff
@@ -52,6 +52,8 @@ folio:
   jpa:
     repository:
       base-packages: org.folio.roles.repository
+  system-user:
+    enabled: true
```
Comment (yauhen-vavilkin) [inline src/main/resources/application.yml:56]: `enabled: false`

File: src/test/resources/application-it.yml [hunk 9]
```diff
@@ -39,3 +39,10 @@ header:
...
+    enabled: false
+    username: test
```
Comment (yauhen-vavilkin) [inline src/test/resources/application-it.yml:46]: remove username and password please

File: src/main/resources/application.yml (hunk 8 above)
Comment (mykolam-epm) [reply src/main/resources/application.yml]: Done

File: src/test/resources/application-it.yml (hunk 9 above)
Comment (mykolam-epm) [reply src/test/resources/application-it.yml:46]: Breaks all integration tests with error `Could not bind properties to 'SystemUserProperties' : prefix=folio.system-user, ignoreInvalidFields=false, ignoreUnknownFields=true`

**Review by OleksiiKuzminov — APPROVED** [review]

**Review by yauhen-vavilkin — APPROVED** [review]

**Pull request merged**